251124 publish single message if batch size is 1

zmstone opened the pull request.

zmstone committed:
fix: ensure to send single messages when batch size is configured to 1

zmstone committed:
fix: ensure client termination and delete ops are atomic

zmstone committed:
test: add inter-op test for Key_Share strategy verifiation

zmstone committed:
ci: add interop test in github actions

thalesmg commented on src/pulsar_producer.erl:
```diff
@@ -887,19 +898,54 @@ do_send_to_pulsar(State0) ->
-            State2 = State1#{requests := Requests, inflight_calls := InflightCalls},
-            State = next_sequence_id(State2),
-            maybe_send_to_pulsar(State)
+            case BatchSize of
```
Minor nit:

```erlang
[_ | _] when BatchSize == 1 ->
    ...;
[_ | _] ->
    ...

thalesmg commented on src/pulsar_producer.erl:
```diff
@@ -887,19 +898,54 @@ do_send_to_pulsar(State0) ->
+    %% Create inflight request for this single message
+    Requests = Requests0#{SequenceId => ?INFLIGHT_REQ(QAckRef, [{From, {Timestamp, [Msg]}}], 1)},
+    InflightCalls = InflightCalls0 + 1,
+    pulsar_metrics:inflight_set(State1, InflightCalls),
```
Nit: might leave bumping the gauge to the clause above so it's done once?

thalesmg commented on src/pulsar_producer.erl:
```diff
@@ -979,10 +1025,17 @@ resend_sent_requests(State) ->
                                                   Msg <- Msgs],
-                                          SequenceId, State),
-                       AccIn#{SequenceId => ?INFLIGHT_REQ(QAckRef, Messages, length(Messages))}
+                       case {_BatchSize, AllMessages} of
```
```suggestion
                       case {BatchSize, AllMessages} of
```

thalesmg commented on src/pulsar_protocol_frame.erl:
```diff
@@ -130,7 +130,17 @@ parse(CmdBin) ->
-            Payloads = parse_batch_message(Payload0, maps:get(num_messages_in_batch, MetadataCmd, 1)),
+            %% If num_messages_in_batch is missing, it's a true single message (non-batch format)
+            %% If present (even if = 1), it's batch format with SingleMessageMetadata headers
+            Payloads = case maps:is_key(num_messages_in_batch, MetadataCmd) of
+                false ->
+                    %% num_messages_in_batch is missing - true single message format, no SingleMessageMetadata
+                    [Payload0];
+                true ->
+                    %% num_messages_in_batch is present - batch format, parse accordingly
+                    NumMessagesInBatch = maps:get(num_messages_in_batch, MetadataCmd),
+                    parse_batch_message(Payload0, NumMessagesInBatch)
+            end,
```
Minor nit:

```suggestion
            Payloads = case maps:find(num_messages_in_batch, MetadataCmd) of
                error ->
                    %% num_messages_in_batch is missing - true single message format, no SingleMessageMetadata
                    [Payload0];
                {ok, NumMessagesInBatch} ->
                    %% num_messages_in_batch is present - batch format, parse accordingly
                    parse_batch_message(Payload0, NumMessagesInBatch)
            end,
```

thalesmg commented on test/pulsar_producer_SUITE.erl:
```diff
@@ -161,3 +171,130 @@ t_port_exit(Config) ->
+    Messages = [
+        #{key => <<"key1">>, value => <<"value1">>},
+        #{key => <<"key2">>, value => <<"value2">>},
+        #{key => <<"key3">>, value => <<"value3">>}
```
Nit: maybe add a message without key to exercise the encoding path where `Key = undefined`?

zmstone committed:
fix: ensure replayq and caller are acked only once

Make sure to acknowledge replayq and caller only once if the
call with multiple messages is split into single messages

zmstone committed:
test: ensure key=undefined is covered

zmstone committed:
test: add case to cover batch_size=1 but caller supplied a batch

zmstone committed:
test: rename escript from .erl to .escript

zmstone committed:
fix: check batch_size in state instead of in sent requests

zmstone committed:
ci: fix python venv
